### ISSUE TITLE
Provide tensor/outer product method 'tf.outer'

### ISSUE THREAD

**Issue by 4d55397500** [tensorflow/tensorflow#17564]
System information
Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux ubuntu 16.04
TensorFlow installed from (source or binary): binary
TensorFlow version (use command below): 1.4.1 gpu
Python version: 3.5.4
Bazel version (if compiling from source):
GCC/Compiler version (if compiling from source):
CUDA/cuDNN version: Cuda 8.0/Cudnn 6.0
GPU model and memory: Titan xp
Exact command to reproduce: no method for outer product

Describe the problem

Tensor (outer product) is the fundamental operation on tensors, but there appears to be no method `tf.outer` in `tensorflow` analogous to `np.outer` in `numpy` to compute the outer product of arbitrary tensors. A google search pulls up these implementation suggestions on stackoverflow: https://stackoverflow.com/questions/33858021/outer-product-in-tensorflow, but these require the dimensions of the tensors to be accessed/known beforehand.

I have the following hack. It would be nice to express this in terms of a pairwise (given associativity) operation `tf.outer`.

The tensor product for an arbitrary collection of tensors can be computed:
```
def tensor_product(*e):
    """ Tensor product of elements """
    if len(e) == 1:
        return e
    elif len(e) == 2:
        a, b = e
        r_a = len(a.get_shape().as_list())
        r_b = len(b.get_shape().as_list())
        s_a = tf.concat([tf.shape(a), tf.constant([1] * r_b)], axis=0)
        s_b = tf.concat([tf.constant([1] * r_a), tf.shape(b)], axis=0)
        a_reshaped = tf.reshape(a, s_a)
        b_reshaped = tf.reshape(b, s_b)
        return a_reshaped * b_reshaped
    prod = e[0]
    for tensor in e[1:]:
        prod = tensor_product(prod, tensor)
    return prod


```

The tensor product allows more elegant expression of rank-2 and greater tensors in a loss function. 
For example here is a diagonal and elliptical (weighted terms) quadratic term:

```
def diagonal_M(batch_size, d):
    """ M_abij = delta_ab delta_ij """
    return tensor_product(tf.diag([1.] * batch_size), tf.diag([1.] * d))

```

```
def biased_diagonal_M(batch_size, d):
    """ M_abij = lambda_i delta_ab delta_ij """
    return tensor_product(tf.diag([1.] * batch_size), tf.diag([5.] + [1.] * (d-1)))

```
ref: https://github.com/4d55397500/quadratic-forms-tensorflow

**Comment by asimshankar** [comment]
@rmlarsen : Any thoughts/suggestions on this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 137 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 153 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 168 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 183 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 198 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by 4d55397500** [comment]
Bump

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by 4d55397500** [comment]
Bump, no interest?

**Comment by swierh** [comment]
This functionality already exists in [tf.tensordot](https://www.tensorflow.org/api_docs/python/tf/tensordot), which works like the [numpy.tensordot](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.tensordot.html) which does state the functionality in the documentation.

Code for outer product is the following: 
```python
tf.tensordot(tensor_a, tensor_b, axes=0)
```
(akin to `np.tensordot(tensor_a, tensor_b, axes=0)`

Documenting the outer product functionality of tensordot and/or creating an alias could fix this issue.

**Comment by 4d55397500** [comment]
thanks @swierh, I appreciate it. Looks like a documentation problem. I may in fact suggest implementing a `tf.outer` operation analogous to `np.outer.` Closing this issue

**Comment by Nyanraltotlapun** [comment]
It not often used, but used nonetheless.
I personally use it for some perceptron error propagation.
If it will be implemented as separate function maybe there is also some optimization possible in contrast of just using tf.tensordot(tensor_a, tensor_b, axes=0) ?

Also numpy and PyTorch have it, so it takes some time for me to find exactly how to do outer product in Tensorflow.
